### PR TITLE
Separate the physical model into a Model class

### PR DESCRIPTION
**Pull request by BoxiLi**
Previously, the Hamiltonian model in `Processor` is saved as partially initialized pulses. This PR separate the Hamiltonian model from pulse and the `Processor` class. It can now be defined independently and includes only the information of the Hamiltonian. The pulses will only be generated after the circuit is loaded.

The main changes are in `processor.py`.  In particular in the initialization. The others are either docs change or internal updates to fit this new structure. A bunch of getter functions are added to support the old way of building the Hamiltonians using the `Processor`. The data is, however, always saved in `Model`.

In principle `ModelProcessor` can be merged into `Processor`, but it is best to wait for another PR.

There are a few FIXME that are used to remind me to revisit some details later in a different PR. They are unrelated to the Hamiltonian model.

@nathanshammah @quantshah 

Issue #98

**TODO**
- [x] Add more tests
- [ ] Check compatibility with existing notebooks.

**Breaking changes**
- The pulse will not be generated before loading the circuit. So inquiring the Hamiltonian operator through pulse will only return an empty list. One should now use e.g. `Processor.get_control("sx0")` for this.
- `Processor.get_operators_labels()` is renamed as `get_latex_str` for clearity.

### CODE REVIEW

File: src/qutip_qip/device/cavityqed.py [hunk 1]
```diff
@@ -255,3 +128,152 @@ def load_circuit(self, qc, schedule_mode="ASAP", compiler=None):
...
+        num_qubits = self.num_qubits
+        return (
+            ["sx" + str(i) for i in range(num_qubits)]
+            + ["sz" + str(i) for i in range(num_qubits)]
+            + ["g" + str(i) for i in range(num_qubits)]
+        )
+
+    def _set_up_controls(self):
+        """
+        Generate the Hamiltonians for the spinchain model and save them in the
```
Comment (quantshah) [inline src/qutip_qip/device/cavityqed.py:204]: Should be "generate the Hamiltonians for the cavity-qed model"

File: doc/source/qip-processor.rst [hunk 2]
```diff
@@ -159,19 +159,19 @@ To let it find the optimal pulses, we need to give the parameters for :func:`~qu
 
 
       # Same parameter for all the gates
-      qc = QubitCircuit(N=1)
+      qc = QubitCircuit(1)
```
Comment (nathanshammah) [inline doc/source/qip-processor.rst:162]:
```suggestion
      qc = QubitCircuit(1)
```
just a question, why did you delete "N=" here and elsewhere?

File: src/qutip_qip/device/cavityqed.py [hunk 3]
```diff
@@ -255,3 +128,152 @@ def load_circuit(self, qc, schedule_mode="ASAP", compiler=None):
         )
         self.global_phase = compiler.global_phase
         return tlist, coeff
+
+
+class CavityQEDModel(Model):
+    """
+    The physical model for a dispersive cavity-QED system.
```
Comment (nathanshammah) [inline src/qutip_qip/device/cavityqed.py:135]:
```suggestion
    The physical model for a dispersive cavity-QED system.
```
shall we change "system" with "processor"? Is there any other detail to add?

File: src/qutip_qip/device/cavityqed.py [hunk 4]
```diff
@@ -255,3 +128,152 @@ def load_circuit(self, qc, schedule_mode="ASAP", compiler=None):
...
+
+    Parameters
+    ----------
+    num_qubits : int
+        The number of component systems.
```
Comment (nathanshammah) [inline src/qutip_qip/device/cavityqed.py:140]:
```suggestion
        The number of component systems.
```
Sounds a bit strange. "The number of qubits"? or "The number of qubits composing the system"?

File: src/qutip_qip/device/cavityqed.py [hunk 5]
```diff
@@ -255,3 +128,152 @@ def load_circuit(self, qc, schedule_mode="ASAP", compiler=None):
...
+    num_levels : int, optional
+        The truncation level of the resonator.
```
Comment (nathanshammah) [inline src/qutip_qip/device/cavityqed.py:142]:
```suggestion
        The truncation level of the Hilbert space for the resonator.
```

File: src/qutip_qip/device/cavityqed.py (hunk 3 above)
Comment (nathanshammah) [reply src/qutip_qip/device/cavityqed.py:135]: Let's add brief details about the model Hamiltonian (should be the Tavis-Cummings Hamiltonian).

File: src/qutip_qip/device/circuitqed.py [hunk 6]
```diff
@@ -27,97 +31,145 @@ class SCQubits(ModelProcessor):
...
+        self._default_compiler = SCQubitsCompiler
+        self.pulse_mode = "continuous"
+
+    def topology_map(self, qc):
+        return to_chain_structure(qc)
+
+
+class SCQubitsModel(Model):
+    """
+    The physical model for superconducting qubits with fixed frequency.
```
Comment (nathanshammah) [inline src/qutip_qip/device/circuitqed.py:64]:
```suggestion
    The physical model for superconducting-qubit model processor with fixed frequency.
```

Comment (nathanshammah) [reply src/qutip_qip/device/circuitqed.py:64]: We should add briefly a bit more details about the Hamiltonian. I propose we write it using the math directive in latex.

File: src/qutip_qip/device/processor.py [hunk 7]
```diff
@@ -1020,3 +1198,120 @@ def _pulse_interpolate(pulse, tlist):
...
+        label_list : list of hashable objects
+            A list of hashable objects each corresponds to
+            an available control Hamiltonian.
+        """
+        return list(self._controls.keys())
+
+    def get_noise(self) -> List[Noise]:
+        """
+        Get a list of :obj:`.Noise` objects.
+        Single qubit relaxation (T1, T2) are not included here.
```
Comment (nathanshammah) [inline src/qutip_qip/device/processor.py:1286]: Why are the local T1 and T2 not included here?

**Review by nathanshammah — CHANGES_REQUESTED** [review]
I left some comments and suggestions. 

Actually the code climate issues seem reasonable and could be actioned upon: besides blank spaces, there is a chunk of code involving pulses that maybe could be refactored in a private function.

Great to see these additions and the thorough testing.

**Comment by BoxiLi** [comment]
@nathanshammah 
>there is a chunk of code involving pulses that maybe could be refactored in a private function.

Which part do you mean?

File: doc/source/qip-processor.rst (hunk 2 above)
Comment (BoxiLi) [reply doc/source/qip-processor.rst:162]: I'm trying to discourage the use of capital `N` as the number of qubits, and eventually replace it with `num_qubits`, to be consistent.

But maybe this should not be included here in this PR, I agree. I can revert it if desired.

File: src/qutip_qip/device/processor.py (hunk 7 above)
Comment (BoxiLi) [reply src/qutip_qip/device/processor.py:1286]: Yeah there is a trade-off. They are not included here because they are defined just as two numbers and will be handled in the processor. I can add them directly here but then whoever implements their own Model will need to do it themselves.

On the other hand, separating them is indeed a bit strange. Now looking at them again, I think it might be better to include also T1 T2 here. Whoever implements a custom model will need to do their own homework, which should not be that difficult if they are already at this stage.

**Comment by BoxiLi** [comment]
@nathanshammah 
>there is a chunk of code involving pulses that maybe could be refactored in a private function.

Which part do you mean?

**Comment by nathanshammah** [comment]
Is this ready for review? The part of code I refer to is the one picked by code climate, 
```
  if isinstance(coeffs, dict):
            pulse_dict = self.get_pulse_dict()
            for pulse_name, value in coeffs.items():
                if pulse_name in pulse_dict:
                    self.pulses[pulse_dict[pulse_name]].coeff = value
```
anyways, this is fixed now.

**Comment by BoxiLi** [comment]
> Is this ready for review?

Yes.

I think it is better to leave a more thorough doc including math equations in a separate PR and get this merged first. We can also update them after the resubmission.

**Review by nathanshammah — APPROVED** [review]
LGTM, thanks @BoxiLi.
